Improvements to cronjob-ldap-group-sync.yml

jkupferer opened the pull request:
#### What is this PR About?

This PR includes improvements to the ldap group sync cronjob template:

* Move ldap group sync bind password to secret
* Enable LDAP TLS with CA certificate
* Change default image from jenkins slave image to openshift3/ose-cli:v3.11
* Add template parameters:
  * LDAP_GROUP_NAME_ATTRIBUTES
  * LDAP_GROUP_MEMBERSHIP_ATTRIBUTES
  * LDAP_GROUP_UID_ATTRIBUTE
  * LDAP_USER_NAME_ATTRIBUTES
  * LDAP_USER_UID_ATTRIBUTE

#### How do we test this?

Apply template and confirm functionality.

Tested on OCP 4.1 labs.opentlc.com cluster with following parameters:

```
oc process -f cronjob-ldap-group-sync.yml \
  -p LDAP_URL=ldap://ipa.shared.example.opentlc.com \
  -p LDAP_BIND_DN=uid=admin,cn=users,cn=accounts,dc=shared,dc=example,dc=opentlc,dc=com \
  -p LDAP_BIND_PASSWORD='******' \
  -p LDAP_GROUPS_FILTER='(&(!(objectClass=mepManagedEntry))(!(cn=trust admins))(!(cn=groups))(!(cn=admins))(!(cn=ipausers))(!(cn=editors))(!(cn=ocp-users))(!(cn=evmgroup*))(!(cn=ipac*)))' \
  -p LDAP_GROUPS_SEARCH_BASE='cn=groups,cn=accounts,dc=shared,dc=example,dc=opentlc,dc=com' \
  -p LDAP_USERS_SEARCH_BASE='cn=users,cn=accounts,dc=shared,dc=example,dc=opentlc,dc=com' \
  -p LDAP_CA_CERT="$(cat ipa-ca.crt)" \
| oc apply -f -
```

cc: @redhat-cop/casl

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -7,44 +7,64 @@ metadata:
+        userUIDAttribute: "${LDAP_USER_UID_ATTRIBUTE}"
+        tolerateMemberNotFoundErrors: true
+        tolerateMemberOutOfScopeErrors: true
+    ldap-ca.crt: "${LDAP_CA_CERT}"
```
Obviously, `ldaps` should be preferred over just `ldap`, but what if `ldap` is used and not `ldaps`? this template now requires it to be `ldaps` (and the user HAS to supply a CA even if `ldap` is used) Could this value just be a "garbage" string if `ldap` is used (instead of `ldap`)? If so, please default a value in the parameters to avoid requiring this to be set for cases where `ldap` is used

jkupferer replied:
I guess I regard using ldap without TLS for any aspect of cluster security as so wildly irresponsible that it would be better to break and make people do some soul-searching about why they don't have at the very least a self-signed certificate.

I have added parameters LDAP_INSECURE and LDAP_CA_CERT_PATH which can be used to disable TLS by setting LDAP_INSECURE=true and LDAP_CA_CERT_PATH=''

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -7,44 +7,64 @@ metadata:
+      kind: "LDAPSyncConfig"
+      apiVersion: "v1"
+      url: "${LDAP_URL}"
+      insecure: false
```
Make this a parameter to set true v.s. false

jkupferer replied:
Done.

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -7,44 +7,64 @@ metadata:
+      bindDN: "${LDAP_BIND_DN}"
+      bindPassword:
+        file: "/etc/secrets/bind_password"
+      ca: "/etc/config/ldap-ca.crt"
```
See other comment re: `ldap` v.s. `ldaps`

jkupferer replied:
Fixed with LDAP_CA_CERT_PATH.

Nitpick... We are technically talking ldap (without TLS) vs. two ways that ldap implements TLS with ldap+starttls and also ldaps. So it isn't really `ldap` vs. `ldaps`.

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -7,44 +7,64 @@ metadata:
     description: "Scheduled Task to Perform LDAP Group Synchronization"
     iconClass: "icon-shadowman"
     tags: "management,cronjob,ldap,group,sync"
+labels:
```
Will labels at this level apply to all objects within this template?

jkupferer replied:
Yes. Double checked just now.

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -62,65 +82,71 @@ objects:
                   - "/bin/bash"
                   - "-c"
-                  - "oc adm groups sync --sync-config=/etc/config/ldap-group-sync.yaml --confirm || :"
+                  - "oc adm groups sync --sync-config=/etc/config/ldap-group-sync.yaml --confirm"
```
This needs to be re-added - i.e.: the ` || : `. Please see this PR for the reason behind this:
https://github.com/redhat-cop/openshift-management/pull/24/files

jkupferer replied:
Hrm... I want to research this more. This doesn't seem like an appropriate fix as it would mask when ldap group sync is indeed failing. It seems the correct solution is to set `spec.backoffLimit` to 0 on the Job (cronjob's `spec.template.spec.backoffLimit`). I've tested this with OCP4 and it works as expected. Still need to check with 3.11.

We may not have had the option of using the backoffLimit when this PR was opened last April.

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -62,65 +82,71 @@ objects:
   groupNames: null
   metadata:
-    name: system:ldap-group-syncers
-    labels:
```
Is there a reason to remove the `labels` here? ... see my question above re: does a "global" one at the top level apply to all objects and that's the reason?

jkupferer replied:
I removed it because it was redundant with the global labels.

oybed reviewed (changes requested):
@jkupferer Thx for this enhancement. Definitely some good stuff, but also some things that needs to be preserved and/or enhanced a bit more - please see inline comments.

oybed commented:
@jkupferer is there anything in your changes that depend on OCP v4.x? I didn't see any, but curious to know if I may have missed something. 

@etsauer saw your comment re: cutting a new release before merging this. [Release v3.11.1](https://github.com/redhat-cop/openshift-management/releases/tag/v3.11.1) is current, so no action is needed as-is, but @jkupferer may need a new release to leverage this after it has been merged. Also, depending on the answer to my question above re: v4.x, we may need to bump the releases for this repo to `v4.x` something...

jkupferer commented:
I've been testing on OCP 4.1. I am going to test on OCP 3.11 today. I don't believe anything here is 4.1 specific.

jkupferer commented:
@oybed Nothing here should be OCP 4.1 specific. I'm going to test on OCP 3.11 today to be sure. In particular I want to check the behavior of the backoffLimit setting in OCP 3.11

jkupferer commented:
Tested on OCP 3.11, everything looks good. Also `spec.template.spec.backoffLimit` of 0 prevents respawn on failure. I also discovered that I need the `--whitelist` option for the GPTE labs, so added `LDAP_GROUPS_WHITELIST` parameter

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -159,6 +188,18 @@ parameters:
+  - name: "LDAP_CA_CERT_PATH"
+    displayName: "LDAP CA Certificate"
+    description: "The LDAP TLS Certificate Authority Certificate Bundle"
+    value: "/etc/config/ldap-ca.crt"
```
Can we please default this to `''`?

jkupferer replied:
I feel defaulting this to empty string will make the template significantly more confusing to use.

If we really want to support insecure configuration as the default then I think we should have separate cronjob group-sync templates.

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -159,6 +188,18 @@ parameters:
+    description: "The name for the secret in which to store the bind password"
+    value: "ldap-bind-password"
+    required: true
+  - name: "LDAP_CA_CERT"
```
Can we default this to `''`?

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -169,6 +210,40 @@ parameters:
+    displayName: "LDAP Insecure"
+    description: "Disable TLS negotiation for LDAP. If set to true, then LDAP_CA_CERT_PATH must be set to empty string."
+    required: true
+    value: "false"
```
To stay backwards compatible, can we please default this to `true`? Also, I'd recommend moving this parameter up, next to the certs params.

oybed commented on jobs/cronjob-ldap-group-sync.yml:
```diff
@@ -62,65 +84,72 @@ objects:
                   - "-c"
-                  - "oc adm groups sync --sync-config=/etc/config/ldap-group-sync.yaml --confirm || :"
+                  - oc adm groups sync --sync-config=/etc/config/ldap-group-sync.yaml --confirm
+                    $([ -s /etc/config/whitelist.txt ] && echo --whitelist=/etc/config/whitelist.txt)
```
I presume you tested and found this to be working, but a quick check with `oc process` of this template shows that `&` gets expanded to `\u0026` ... again, this may not be a problem, but wanted to ensure this is expected behavior

```
> oc process -f jobs/cronjob-ldap-group-sync.yml ...  | grep 'oc adm groups'
 "oc adm groups sync --sync-config=/etc/config/ldap-group-sync.yaml --confirm $([ -s /etc/config/whitelist.txt ] \u0026\u0026 echo --whitelist=/etc/config/whitelist.txt)"
```

jkupferer replied:
Yes, I tested and all is well. I'm not sure why it gets expanded like this, but in the resulting cronjob it is back to "&"

oybed reviewed (changes requested):
Per convo in chat, added a few more comments inline + a new observation/question.

jkupferer committed:
Add secure ldap group sync job template

Also include improvements to the insecure ldap group sync template:

* Move ldap group sync bind password to secret
* Allow job failure without retry using job backoffLimit=0
* Change default image from jenkins slave image to openshift3/ose-cli:v3.11
* Add template parameters:
 - LDAP_GROUP_NAME_ATTRIBUTES
 - LDAP_GROUP_MEMBERSHIP_ATTRIBUTES
 - LDAP_GROUP_UID_ATTRIBUTE
 - LDAP_GROUPS_WHITELIST
 - LDAP_INSECURE
 - LDAP_USER_NAME_ATTRIBUTES
 - LDAP_USER_UID_ATTRIBUTE

oybed reviewed (approved):
LGTM
